src/lib.rs

```diff
@@ -1,14 +1,207 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::{cmp::Ordering, hash};
+
+pub use proc_macros::grouped_ordering;
+
+pub trait GroupedOrdering {
+    type Group: hash::Hash + Eq;
+
+    fn compare(&self, a: &Self::Group, b: &Self::Group) -> Ordering;
+}
+
+pub trait GroupedOrderable<TGroupedOrdering: GroupedOrdering> {
+    fn map_to_grouped_ordering(&self) -> TGroupedOrdering::Group;
+}
+
+pub trait VecExt {
+    type Item;
+
+    fn sort_by_grouped_ordering<TGroupedOrdering>(&mut self, grouped_ordering: &TGroupedOrdering)
+    where
+        TGroupedOrdering: GroupedOrdering,
+        Self::Item: GroupedOrderable<TGroupedOrdering>;
+}
+
+impl<T> VecExt for Vec<T> {
+    type Item = T;
+
+    fn sort_by_grouped_ordering<TGroupedOrdering>(&mut self, grouped_ordering: &TGroupedOrdering)
+    where
+        TGroupedOrdering: GroupedOrdering,
+        Self::Item: GroupedOrderable<TGroupedOrdering>,
+    {
+        self.sort_by(|a, b| {
+            grouped_ordering.compare(&a.map_to_grouped_ordering(), &b.map_to_grouped_ordering())
+        });
+    }
 }
 
 #[cfg(test)]
 mod tests {
+    use std::collections::HashMap;
+
+    use speculoos::prelude::*;
+
     use super::*;
+    use proc_macros::grouped_ordering_crate_internal as grouped_ordering;
+
+    #[test]
+    fn test_manual_impl_vec_sort() {
+        #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
```

Comment:
I used this as basically a "template" for what the proc macro should generate so I'll comment on some implementation stuff here